File: jumble_solver.java
import java.util.*;
import java.io.*;


public class jumble_solver {

    public static void main(String[] args) throws IOException {
        if(args.length < 1) {
            System.out.println("usage: java jumble_solver DICTIONARY_FILE [DICTIONARY_FILE] ...");
            System.exit(1);
        }

        HashMap<String,TreeSet<String>> sortedToOrigs = new HashMap<String,TreeSet<String>>();

        for(String dictFileName : args) {
            BufferedReader dictFileReader = new BufferedReader(new FileReader(dictFileName));
            String origWord;

            while((origWord = dictFileReader.readLine()) != null) {
                origWord = origWord.toLowerCase();
                String sortedWord = sorted(origWord);
                TreeSet<String> origs = sortedToOrigs.get(sortedWord);

                if(origs == null) {
                    origs = new TreeSet<String>();
                    sortedToOrigs.put(sortedWord, origs);
                }

                origs.add(origWord);
            }
        }

        BufferedReader stdinReader = new BufferedReader(new InputStreamReader(System.in));
        String origWord;

        System.out.print("$ ");
        System.out.flush();

        while((origWord = stdinReader.readLine()) != null) {
            origWord = origWord.toLowerCase();

            if(origWord.length() > 0) {
                String wordSorted = sorted(origWord);

                if(sortedToOrigs.containsKey(wordSorted)) {
                    TreeSet<String> anagrams = sortedToOrigs.get(wordSorted);

                    for(String word : anagrams) {
                        System.out.print(word + " ");
                    }

                    System.out.println();
                }
                else{
                    System.out.println("no anagram in dictionary");
                }
            }

            System.out.print("$ ");
            System.out.flush();
        }

        System.out.println();
    }


    public static String sorted(String str) {
        char[] letters = str.toCharArray();
        Arrays.sort(letters);
        return new String(letters);
    }
}

File: jumble_solver.py
#! /usr/bin/env python

import sys
import fileinput


if len(sys.argv) < 2:
    print "usage: jumble_solver.py DICTIONARY_FILE [DICTIONARY_FILE] ..."
    exit(1)

sortedToOrigs = {};

for line in fileinput.input():
    origWord = line.strip().lower()
    sortedWord = "".join(sorted(origWord))

    if sortedToOrigs.has_key(sortedWord):
        origs = sortedToOrigs[sortedWord].add(origWord)
    else:
        sortedToOrigs[sortedWord] = set([origWord])

sys.stdout.write('$ ')

while True:
    line = sys.stdin.readline()

    if len(line) == 0:
        break

    sortedWord = "".join(sorted(line.strip().lower()))

    if sortedToOrigs.has_key(sortedWord):
        print " ".join(sortedToOrigs[sortedWord])
    else:
        print "no anagram in dictionary"

    sys.stdout.write('$ ')

print


